src/shared/util.js

```diff
@@ -77,7 +77,7 @@ export function isValidArrayIndex (val: any): boolean {
 export function toString (val: any): string {
   return val == null
     ? ''
-    : typeof val === 'object'
+    : typeof val === 'object' && (!isPlainObject(val) || val.toString === _toString)
```

Comment:
I think this function can really use early returns now.

Reply:
As you want, I kept the original style to make as few changes as possible and let the diff easier to read, but it'll be easy to refactor.